EventFilter/Phase2TrackerRawToDigi/plugins/DTHDAQToFEDRawDataConverter.cc

```diff
@@ -0,0 +1,205 @@
+      return value;
+    };
+
+    uint16_t version = static_cast<uint16_t>(readLittleEndian(&buffer[startIdx], 2));
+    startIdx += 2;
+    uint32_t sourceId = static_cast<uint32_t>(readLittleEndian(&buffer[startIdx], 4));
+    startIdx += 4;
+    uint32_t runNumber = static_cast<uint32_t>(readLittleEndian(&buffer[startIdx], 4));
+    startIdx += 4;
+    uint32_t orbitNumber = static_cast<uint32_t>(readLittleEndian(&buffer[startIdx], 4));
+    startIdx += 4;
+
+    uint32_t eventCountReserved = static_cast<uint32_t>(readLittleEndian(&buffer[startIdx], 4));
+    uint16_t eventCount = eventCountReserved & 0xFFF;
+    startIdx += 4;
+    uint32_t packetWordCount = static_cast<uint32_t>(readLittleEndian(&buffer[startIdx], 4));
+    startIdx += 4;
+    uint32_t flags = static_cast<uint32_t>(readLittleEndian(&buffer[startIdx], 4));
+    startIdx += 4;
+    uint32_t checksum = static_cast<uint32_t>(readLittleEndian(&buffer[startIdx], 4));
+    startIdx += 4;
+
+    std::cout << "Version: " << version << "\nSource ID: " << sourceId << "\nRun Number: " << runNumber
+              << "\nOrbit Number: " << orbitNumber << "\nEvent Count: " << eventCount
+              << "\nPacket Word Count: " << packetWordCount << "\nFlags: " << flags
+              << "\nChecksum: 0x" << std::hex << checksum << std::dec << "\n";
+
+    // Reverse parse fragments within this orbit
+    reverseParseFragments(buffer, startIdx, packetWordCount, eventCount);
+  }
+}
+
+void DTHDAQToFEDRawDataConverter::reverseParseFragments(const std::vector<char>& buffer, size_t startIdx, uint32_t packetWordCount, uint16_t eventCount) {
+  size_t bufferSize = buffer.size();
+
+  size_t index = startIdx + (packetWordCount * 16);  // Calculate where the orbit payload ends
+
+  std::cout << "Starting reverse parsing from byte offset: " << index << std::endl;
+
+  auto readLittleEndian = [&](const char* data, size_t size) -> uint64_t {
+    uint64_t value = 0;
+    for (size_t i = 0; i < size; ++i) {
+      value |= (static_cast<uint64_t>(static_cast<unsigned char>(data[i])) << (8 * i));
+    }
+    return value;
+  };
+
+  for (int frag = eventCount - 1; frag >= 0; --frag) {
+    if (index < 16) {
+      std::cerr << "Not enough data for fragment trailer of fragment " << frag + 1 << std::endl;
+      return;
+    }
+    index -= 16;
+
+    uint8_t markerF = static_cast<uint8_t>(buffer[index]);
+    uint8_t markerH = static_cast<uint8_t>(buffer[index + 1]);
+    if (markerF != 0x48 || markerH != 0x46) {
+      std::cerr << "Invalid Fragment Trailer marker in fragment " << frag + 1 << ": 0x"
+                << std::hex << static_cast<int>(markerF) << " 0x" << static_cast<int>(markerH) << std::dec << std::endl;
+      return;
+    }
+
+    uint16_t fragFlags = static_cast<uint16_t>(readLittleEndian(&buffer[index + 2], 2));
+    uint32_t fragSize = static_cast<uint32_t>(readLittleEndian(&buffer[index + 4], 4));
+    uint64_t eventId = readLittleEndian(&buffer[index + 8], 8) & 0xFFFFFFFFFFF;
```

Comment:
You have lots of magic numbers representing bit field widths. Turn these into named constants. e.g. constexpr unsigned int eventBitWidth = 4; and put them in a separate file. But tak also to Andrew & Sara, since they must use many of the same constants, so you shoulld be able to share the same file.

Reply:
I addressed this and changed them to named constants. For now we can have them in the same script as the rest of the code, but I am working on putting them in a separate header file.